Riot/AppDelegate.m

```diff
@@ -3288,4 +3309,109 @@ - (void)disableNoVoIPOnMatrixSession:(MXSession*)mxSession
+         [self checkPendingRoomKeyRequestsInSession:mxSession];
+     }];
+
+    roomKeyRequestCancellationObserver  =
+    [[NSNotificationCenter defaultCenter] addObserverForName:kMXCryptoRoomKeyRequestCancellationNotification
+                                                      object:mxSession.crypto
+                                                       queue:[NSOperationQueue mainQueue]
+                                                  usingBlock:^(NSNotification *notif)
+     {
+         [self checkPendingRoomKeyRequestsInSession:mxSession];
+     }];
+}
+
+- (void)disableRoomKeyRequestObserver:(MXSession*)mxSession
+{
+    if (roomKeyRequestObserver)
+    {
+        [[NSNotificationCenter defaultCenter] removeObserver:roomKeyRequestObserver];
+        roomKeyRequestObserver = nil;
+    }
+
+    if (roomKeyRequestCancellationObserver)
+    {
+        [[NSNotificationCenter defaultCenter] removeObserver:roomKeyRequestCancellationObserver];
+        roomKeyRequestCancellationObserver = nil;
+    }
+}
+
+// Check if a key share dialog must be displayed for the given session
+- (void)checkPendingRoomKeyRequestsInSession:(MXSession*)mxSession
+{
+    if ([UIApplication sharedApplication].applicationState != UIApplicationStateActive)
+    {
+        NSLog(@"[AppDelegate] checkPendingRoomKeyRequestsInSession called while the app is not active. Ignore it.");
+        return;
+    }
+
+    [mxSession.crypto pendingKeyRequests:^(MXUsersDevicesMap<NSArray<MXIncomingRoomKeyRequest *> *> *pendingKeyRequests) {
+
+        NSLog(@"[AppDelegate] checkPendingRoomKeyRequestsInSession: pendingKeyRequests.count: %@. Already displayed: %@",
+              @(pendingKeyRequests.count),
+              roomKeyRequestViewController ? @"YES" : @"NO");
+
+        if (roomKeyRequestViewController)
+        {
+            // Check if the current RoomKeyRequestViewController is still valid
+            MXSession *currentMXSession = roomKeyRequestViewController.mxSession;
+            NSString *currentUser = roomKeyRequestViewController.device.userId;
+            NSString *currentDevice = roomKeyRequestViewController.device.deviceId;
+
+            NSArray<MXIncomingRoomKeyRequest *> *currentPendingRequest = [pendingKeyRequests objectForDevice:currentDevice forUser:currentUser];
+
+            if (currentMXSession == mxSession && currentPendingRequest.count == 0)
+            {
+                NSLog(@"[AppDelegate] checkPendingRoomKeyRequestsInSession: Cancel current dialog");
+
+                // The key request has been probably cancelled, remove the popup
+                [roomKeyRequestViewController hide];
+                roomKeyRequestViewController = nil;
+            }
+        }
+
+        if (!roomKeyRequestViewController && pendingKeyRequests.count)
+        {
+            // Pick the first coming user/device pair
+            NSString *user = pendingKeyRequests.userIds[0];
+            NSString *device = [pendingKeyRequests deviceIdsForUser:user][0];
```

Comment:
You should use `firstObject` in case `[pendingKeyRequests deviceIdsForUser:user]` is empty.